File: typescript/src/http.ts
import type { ApiResponse } from "./types"

export class SentroyError extends Error {
  constructor(
    public readonly statusCode: number,
    public readonly body: unknown,
    message?: string,
  ) {
    super(message || `Sentroy API error (${statusCode})`)
    this.name = "SentroyError"
  }
}

export class HttpClient {
  private baseUrl: string
  /**
   * Bearer access token. Tanımsızsa Authorization header eklenmez ve
   * `credentials: "include"` ile cookie auth kullanılır — bu mod Sentroy
   * site içinden (sentroy.com / mail.sentroy.com / storage.sentroy.com)
   * MediaManager gibi React component'leri kullanırken caller'ın access
   * token üretmesine gerek bırakmaz; mevcut session cookie'si yeter.
   */
  private token: string | undefined
  private timeout: number

  constructor(baseUrl: string, token: string | undefined, timeout = 30_000) {
    this.baseUrl = baseUrl.replace(/\/+$/, "")
    this.token = token
    this.timeout = timeout
  }

  private buildUrl(path: string, query?: Record<string, unknown>): string {
    const url = new URL(`${this.baseUrl}${path}`)
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined && value !== null) {
          url.searchParams.set(key, String(value))
        }
      }
    }
    return url.toString()
  }

  private authHeaders(): Record<string, string> {
    return this.token ? { Authorization: `Bearer ${this.token}` } : {}
  }

  /** Cookie modunda credentials include; bearer modunda omit. */
  private get fetchCredentials(): RequestCredentials {
    return this.token ? "omit" : "include"
  }

  /**
   * Raw fetch — JSON parse etmez, Response döner. Binary endpoint'ler
   * (media download) için kullanılır. Authorization header eklenir veya
   * cookie auth ile çalışılır.
   */
  async fetchRaw(
    path: string,
    init?: { method?: string; query?: Record<string, unknown> },
  ): Promise<Response> {
    const url = this.buildUrl(path, init?.query)
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeout)
    try {
      return await fetch(url, {
        method: init?.method || "GET",
        headers: this.authHeaders(),
        credentials: this.fetchCredentials,
        signal: controller.signal,
      })
    } finally {
      clearTimeout(timer)
    }
  }

  private async request<T>(
    method: string,
    path: string,
    options?: {
      query?: Record<string, unknown>
      body?: unknown
    },
  ): Promise<T> {
    const url = this.buildUrl(path, options?.query)
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeout)

    try {
      const headers: Record<string, string> = { ...this.authHeaders() }
      if (options?.body) {
        headers["Content-Type"] = "application/json"
      }

      const res = await fetch(url, {
        method,
        headers,
        body: options?.body ? JSON.stringify(options.body) : undefined,
        credentials: this.fetchCredentials,
        signal: controller.signal,
      })

      const json = (await res.json()) as ApiResponse<T>

      if (!res.ok) {
        throw new SentroyError(
          res.status,
          json,
          json.error || `Request failed with status ${res.status}`,
        )
      }

      return json.data
    } finally {
      clearTimeout(timer)
    }
  }

  async get<T>(path: string, query?: Record<string, unknown>): Promise<T> {
    return this.request<T>("GET", path, { query })
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>("POST", path, { body })
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>("PUT", path, { body })
  }

  async patch<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>("PATCH", path, { body })
  }

  async del<T>(path: string, query?: Record<string, unknown>): Promise<T> {
    return this.request<T>("DELETE", path, { query })
  }

  /**
   * DELETE with a JSON body. Most endpoints take an id in the path, but a
   * few collection-membership endpoints (audience list members) carry the
   * member id in the body so a single route can address both members of
   * the parent. Fetch + undici both allow DELETE bodies — keep separate
   * from `del` so the common case stays query-only.
   */
  async delWithBody<T>(path: string, body: unknown): Promise<T> {
    return this.request<T>("DELETE", path, { body })
  }

  /**
   * Multipart upload. Body is a FormData; we deliberately do not set
   * Content-Type so the runtime (browser or Node undici) writes the
   * correct `multipart/form-data; boundary=...` header.
   */
  async postForm<T>(path: string, form: FormData): Promise<T> {
    const url = this.buildUrl(path)
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeout)
    try {
      const res = await fetch(url, {
        method: "POST",
        headers: this.authHeaders(),
        credentials: this.fetchCredentials,
        body: form,
        signal: controller.signal,
      })
      const json = (await res.json()) as { data: T; error?: string }
      if (!res.ok) {
        throw new SentroyError(
          res.status,
          json,
          json.error || `Upload failed with status ${res.status}`,
        )
      }
      return json.data
    } finally {
      clearTimeout(timer)
    }
  }

  /**
   * XHR-tabanlı multipart upload — `fetch` progress event veremiyor, XHR
   * `upload.onprogress` ile bytes loaded/total event'i tetikler. Caller
   * `onProgress` ve `signal` (AbortController) verir.
   *
   * Cookie auth (`token` undefined) için `withCredentials = true`. Timeout
   * uygulanmaz (büyük dosyalar için 30sn'de patlayıp kaybolmasın); caller
   * istiyorsa `signal` ile kendi timeout'unu kurar.
   */
  async postFormWithProgress<T>(
    path: string,
    form: FormData,
    opts: {
      onProgress?: (loaded: number, total: number) => void
      signal?: AbortSignal
    } = {},
  ): Promise<T> {
    const url = this.buildUrl(path)
    return new Promise<T>((resolve, reject) => {
      const xhr = new XMLHttpRequest()
      xhr.open("POST", url, true)
      if (this.token) {
        xhr.setRequestHeader("Authorization", `Bearer ${this.token}`)
      } else {
        xhr.withCredentials = true
      }

      if (opts.signal) {
        if (opts.signal.aborted) {
          reject(new SentroyError(0, null, "Upload aborted"))
          return
        }
        opts.signal.addEventListener(
          "abort",
          () => {
            try {
              xhr.abort()
            } catch {
              /* noop */
            }
            reject(new SentroyError(0, null, "Upload aborted"))
          },
          { once: true },
        )
      }

      if (opts.onProgress) {
        xhr.upload.addEventListener("progress", (e) => {
          if (e.lengthComputable) {
            opts.onProgress!(e.loaded, e.total)
          }
        })
      }

      xhr.addEventListener("load", () => {
        let json: { data?: T; error?: string } = {}
        try {
          json = JSON.parse(xhr.responseText) as { data?: T; error?: string }
        } catch {
          /* fall through with empty object */
        }
        if (xhr.status >= 200 && xhr.status < 300 && json.data !== undefined) {
          resolve(json.data)
        } else {
          reject(
            new SentroyError(
              xhr.status,
              json,
              json.error || `Upload failed with status ${xhr.status}`,
            ),
          )
        }
      })

      xhr.addEventListener("error", () =>
        reject(new SentroyError(0, null, "Upload network error")),
      )
      xhr.addEventListener("timeout", () =>
        reject(new SentroyError(0, null, "Upload timed out")),
      )

      xhr.send(form)
    })
  }
}

File: go/http.go
package sentroy

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// SentroyError is returned when the API responds with a non-2xx status code.
type SentroyError struct {
	StatusCode int
	Body       string
	Message    string
}

func (e *SentroyError) Error() string {
	if e.Message != "" {
		return e.Message
	}
	return fmt.Sprintf("Sentroy API error (%d)", e.StatusCode)
}

// httpClient wraps net/http.Client and handles auth, serialisation, and errors.
type httpClient struct {
	baseURL string
	token   string
	client  *http.Client
}

func newHTTPClient(baseURL, token string, timeout time.Duration) *httpClient {
	return &httpClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		token:   token,
		client:  &http.Client{Timeout: timeout},
	}
}

func (h *httpClient) buildURL(path string, query map[string]string) (string, error) {
	u, err := url.Parse(h.baseURL + path)
	if err != nil {
		return "", err
	}
	if len(query) > 0 {
		q := u.Query()
		for k, v := range query {
			q.Set(k, v)
		}
		u.RawQuery = q.Encode()
	}
	return u.String(), nil
}

// doRequest performs an HTTP request, unwraps the {"data": T} envelope, and
// returns the typed result. It is a generic helper used by every service.
func doRequest[T any](h *httpClient, method, path string, query map[string]string, body interface{}) (T, error) {
	var zero T

	fullURL, err := h.buildURL(path, query)
	if err != nil {
		return zero, err
	}

	var reqBody io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return zero, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewReader(data)
	}

	req, err := http.NewRequest(method, fullURL, reqBody)
	if err != nil {
		return zero, err
	}

	req.Header.Set("Authorization", "Bearer "+h.token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := h.client.Do(req)
	if err != nil {
		return zero, err
	}
	defer resp.Body.Close()

	rawBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return zero, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var envelope apiResponse[json.RawMessage]
		msg := fmt.Sprintf("Request failed with status %d", resp.StatusCode)
		if json.Unmarshal(rawBody, &envelope) == nil && envelope.Error != "" {
			msg = envelope.Error
		}
		return zero, &SentroyError{
			StatusCode: resp.StatusCode,
			Body:       string(rawBody),
			Message:    msg,
		}
	}

	var envelope apiResponse[T]
	if err := json.Unmarshal(rawBody, &envelope); err != nil {
		return zero, fmt.Errorf("failed to decode response: %w", err)
	}

	return envelope.Data, nil
}

// multipartField is one piece of a multipart/form-data request. Use text for
// string values and file (with filename + content reader) for a file part.
type multipartField struct {
	name     string
	text     string
	filename string
	reader   io.Reader
}

// postMultipart uploads a multipart/form-data body and unwraps the envelope.
// Used by media upload; callers build []multipartField explicitly so we don't
// hide the MIME boundary handling.
func postMultipart[T any](h *httpClient, path string, fields []multipartField) (T, error) {
	var zero T

	fullURL, err := h.buildURL(path, nil)
	if err != nil {
		return zero, err
	}

	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)
	for _, f := range fields {
		if f.reader != nil {
			part, err := writer.CreateFormFile(f.name, f.filename)
			if err != nil {
				return zero, fmt.Errorf("failed to create form file: %w", err)
			}
			if _, err := io.Copy(part, f.reader); err != nil {
				return zero, fmt.Errorf("failed to write file part: %w", err)
			}
		} else {
			if err := writer.WriteField(f.name, f.text); err != nil {
				return zero, fmt.Errorf("failed to write field: %w", err)
			}
		}
	}
	if err := writer.Close(); err != nil {
		return zero, fmt.Errorf("failed to close multipart writer: %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, fullURL, body)
	if err != nil {
		return zero, err
	}
	req.Header.Set("Authorization", "Bearer "+h.token)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	resp, err := h.client.Do(req)
	if err != nil {
		return zero, err
	}
	defer resp.Body.Close()

	rawBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return zero, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		var envelope apiResponse[json.RawMessage]
		msg := fmt.Sprintf("Upload failed with status %d", resp.StatusCode)
		if json.Unmarshal(rawBody, &envelope) == nil && envelope.Error != "" {
			msg = envelope.Error
		}
		return zero, &SentroyError{
			StatusCode: resp.StatusCode,
			Body:       string(rawBody),
			Message:    msg,
		}
	}

	var envelope apiResponse[T]
	if err := json.Unmarshal(rawBody, &envelope); err != nil {
		return zero, fmt.Errorf("failed to decode response: %w", err)
	}
	return envelope.Data, nil
}

// fetchRaw issues a GET request and returns the raw response body bytes,
// without the {"data": T} envelope. Used by binary endpoints like media
// download.
func (h *httpClient) fetchRaw(path string, query map[string]string) ([]byte, string, error) {
	fullURL, err := h.buildURL(path, query)
	if err != nil {
		return nil, "", err
	}
	req, err := http.NewRequest(http.MethodGet, fullURL, nil)
	if err != nil {
		return nil, "", err
	}
	req.Header.Set("Authorization", "Bearer "+h.token)

	resp, err := h.client.Do(req)
	if err != nil {
		return nil, "", err
	}
	defer resp.Body.Close()

	rawBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, "", fmt.Errorf("failed to read response body: %w", err)
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, "", &SentroyError{
			StatusCode: resp.StatusCode,
			Body:       string(rawBody),
			Message:    fmt.Sprintf("Download failed with status %d", resp.StatusCode),
		}
	}
	return rawBody, resp.Header.Get("Content-Type"), nil
}
